Dockerfile

```diff
@@ -1,33 +1,58 @@
-FROM ruby:2.7.7-alpine3.16
+######################
+##### Base Stage #####
+######################
 
-# Install runtime dependencies
-RUN apk update && apk upgrade && apk add --no-cache bash build-base git nodejs npm tzdata postgresql-dev
+FROM ruby:2.7.7-alpine3.16 AS base
 
-# Copy all required files
+# Create and define work directory
 WORKDIR /opt/killswitch
+
+# Install OS dependencies
+RUN apk --update --no-cache add nodejs tzdata libpq && \
+  gem update --system 3.0.8 && gem update --system
+
+#######################
+##### Build Stage #####
+#######################
+
+FROM base AS build
+
+# Copy only the minimal required files
 COPY . /opt/killswitch/
-RUN mkdir -p /opt/killswitch/tmp
 
-# Update Ruby-provided Bundler version (via rubygems update)
-RUN gem update --system 3.0.8 && gem update --system
+# Update system and install dependencies
+RUN apk --update --no-cache add --virtual build-dependencies build-base git nodejs npm postgresql-dev && \
+  bundle config set --local without 'development test' && \
+  bundle install && \
+  bundle binstubs --all && \
+  npm set progress=false && \
+  npm install --silent --production
+
+# Pre-compile assets
+RUN SECRET_KEY_BASE=__UNUSED_BUT_REQUIRED__ RAILS_ENV=production bundle exec rake assets:precompile
 
-# Install Ruby dependencies
-RUN bundle install --binstubs --without development test
+#########################
+##### Release Stage #####
+#########################
+
+FROM base AS release
+
+COPY . .
+COPY --from=build /usr/local/bundle/ /usr/local/bundle/
+COPY --from=build /opt/killswitch/public/assets/ /opt/killswitch/public/assets/
+COPY --from=build /opt/killswitch/node_modules/ /opt/killswitch/node_modules/
 
 # Copy entrypoint script
 COPY scripts/docker-entrypoint.sh /usr/local/bin
 RUN chmod a+x /usr/local/bin/docker-entrypoint.sh
 
 # Create user
-RUN adduser -D killswitch
-RUN chown -R killswitch: /opt/killswitch
-USER killswitch
+RUN adduser -D -h /opt/killswitch -u 5000 killswitch && \
+  chown -R killswitch: /opt/killswitch
```

Comment:
Is there a reason to join `adduser` and `chown` in a single `RUN` command here?

Reply:
Each `RUN` instructions create an additional layer. Piping multiple commands to keep the image size as small as possible.

Reference: https://docs.docker.com/develop/develop-images/dockerfile_best-practices/#minimize-the-number-of-layers